File: api-express/src/dal/cliente-dal.js
import prisma from "../configs/prisma-client.js";
import AppMessage from "../utils/app-message.js";

export const create = async (payload) => {
    const clienteExists = await prisma.cliente.findFirst({
        where: {
            OR: [
                {
                    email: {
                        equals: payload.email,
                        mode: 'insensitive',
                    },
                },
                {
                    cpf: {
                        equals: payload.cpf,
                        mode: 'insensitive',
                    },
                },
                {
                    cnpj: {
                        equals: payload.cnpj,
                        mode: 'insensitive',
                    },
                }
            ]
        },
        select: {
            email: true,
            cnpj: true,
            cpf: true,
        }
    });

    if (clienteExists) {
        const msg = payload.email == clienteExists.email ? 'Cliente com email já cadastrado'
                : payload.cnpj == clienteExists.cnpj && payload.cnpj != undefined ? 'Cliente com CNPJ já cadastrado'
                : payload.cpf == clienteExists.cpf && payload.cpf != undefined ? 'Cliente com CPF já cadastrado'
                : '';

        throw new AppMessage(msg, 400);
    }

    const cliente = await prisma.cliente.create({
        data: payload,
    });

    await prisma.$disconnect();

    return cliente;
};

export const getByEmail = async (email) => {
    const cliente = await prisma.cliente.findFirst({
        where: {
            email,
        },
    });

    if (!cliente) {
        throw new AppMessage('Cliente não encontrado', 404);
    }

    await prisma.$disconnect();

    return cliente;
};

File: api-actix/src/dal/cliente_dal.rs
use diesel::prelude::*;
use crate::db::DbPool;
use crate::models::cliente::{Cliente, NewCliente};
use crate::utils::app_message::AppMessage;

pub struct ClienteDal;

impl ClienteDal {
    pub async fn create(pool: &DbPool, payload: NewCliente) -> Result<Cliente, AppMessage> {
        use crate::schema::clientes::dsl::*;
        use diesel::result::{Error as DieselError, DatabaseErrorKind};

        let pool_clone = pool.clone();
        let payload_owned = payload;

        tokio::task::spawn_blocking(move || {
            let mut connection = pool_clone.get()
                .map_err(|e| AppMessage::new(&format!("Database connection error: {}", e), 500))?;

            match diesel::insert_into(clientes)
                .values(&payload_owned)
                .get_result::<Cliente>(&mut connection)
            {
                Ok(cliente) => Ok(cliente),
                Err(DieselError::DatabaseError(DatabaseErrorKind::UniqueViolation, info)) => {
                    let error_message = if let Some(constraint) = info.constraint_name() {
                        match constraint {
                            "clientes_email_key" | "idx_clientes_email" => "Cliente com email já cadastrado",
                            "clientes_cpf_key" | "idx_clientes_cpf" => "Cliente com CPF já cadastrado",
                            "clientes_cnpj_key" | "idx_clientes_cnpj" => "Cliente com CNPJ já cadastrado",
                            "clientes_telefone_key" | "idx_clientes_telefone" => "Cliente com telefone já cadastrado",
                            _ => "Cliente já cadastrado com esses dados"
                        }
                    } else {
                        let details = info.message();
                        if details.contains("email") {
                            "Cliente com email já cadastrado"
                        } else if details.contains("cpf") {
                            "Cliente com CPF já cadastrado"
                        } else if details.contains("cnpj") {
                            "Cliente com CNPJ já cadastrado"
                        } else if details.contains("telefone") {
                            "Cliente com telefone já cadastrado"
                        } else {
                            "Cliente já cadastrado"
                        }
                    };

                    Err(AppMessage::new(error_message, 400))
                },
                Err(e) => Err(AppMessage::new(&format!("Erro ao criar cliente: {}", e), 500))
            }
        }).await
            .map_err(|e| AppMessage::new(&format!("Task error: {}", e), 500))?
    }

    pub async fn get_by_email(pool: &DbPool, email_param: &str) -> Result<Cliente, AppMessage> {
        use crate::schema::clientes::dsl::*;

        let pool_clone = pool.clone();
        let email_owned = email_param.to_string();

        tokio::task::spawn_blocking(move || {
            let mut connection = pool_clone.get()
                .map_err(|e| AppMessage::new(&format!("Database connection error: {}", e), 500))?;

            let cliente = clientes
                .filter(email.eq(&email_owned))
                .first::<Cliente>(&mut connection)
                .optional()
                .map_err(|e| AppMessage::new(&format!("Erro ao buscar cliente: {}", e), 500))?;

            match cliente {
                Some(c) => Ok(c),
                None => Err(AppMessage::new("Cliente não encontrado", 404)),
            }
        }).await
            .map_err(|e| AppMessage::new(&format!("Task error: {}", e), 500))?
    }
}